durabletask/internal/grpc_interceptor.py

```diff
@@ -18,76 +18,44 @@ class _ClientCallDetails(
     """
     pass
 
-def _intercept_call(
-            client_call_details: _ClientCallDetails, metadata: List[Tuple[str, str]]) -> _ClientCallDetails:
-    """Internal intercept_call implementation which adds metadata to grpc metadata in the RPC
-        call details."""
-    new_metadata = []
-    new_client_call_details = client_call_details
-    if client_call_details.metadata is not None:
-        new_metadata = list(metadata)
-    if metadata is not None:
-        new_metadata.extend(metadata)
-        new_client_call_details = _ClientCallDetails(
-            client_call_details.method,
-            client_call_details.timeout,
-            metadata=new_metadata,
-            credentials=client_call_details.credentials,
-            wait_for_ready=client_call_details.wait_for_ready,
-            compression=client_call_details.compression,
-        )
-    return new_client_call_details
 
-class UnaryUnaryClientInterceptorImpl (grpc.UnaryUnaryClientInterceptor):
-    """The class implements a UnaryUnaryClientInterceptor from grpc to add an interceptor to add
-    additional headers to all calls as needed."""
+class DefaultClientInterceptorImpl (
+    grpc.UnaryUnaryClientInterceptor, grpc.UnaryStreamClientInterceptor,
+    grpc.StreamUnaryClientInterceptor, grpc.StreamStreamClientInterceptor):
+    """The class implements a UnaryUnaryClientInterceptor, UnaryStreamClientInterceptor,
+    StreamUnaryClientInterceptor and StreamStreamClientInterceptor from grpc to add an 
+    interceptor to add additional headers to all calls as needed."""
 
     def __init__(self, metadata: List[Tuple[str, str]]):
         super().__init__()
         self._metadata = metadata
 
-    def intercept_unary_unary(self, continuation, client_call_details, request):
-        new_client_call_details = _intercept_call(client_call_details, self._metadata)
-        return continuation(new_client_call_details, request)
+    def _intercept_call(
+                self, client_call_details: _ClientCallDetails) -> grpc.ClientCallDetails:
+        """Internal intercept_call implementation which adds metadata to grpc metadata in the RPC
+            call details."""
+        metadata = []
```

Comment:
Let's move this into an `else:` block so that we don't make any unnecessary allocations.

Reply:
I moved that to an `else` block and also moved creation of _ClientCallDetails, only if passed metadata is not None.